src/server/opc_tcp_async.cpp

```diff
@@ -301,28 +312,54 @@ namespace
   void OpcTcpServer::Listen()
   {
     std::clog << "opc_tcp_async| Running server." << std::endl;
+
+    std::cout << "opc_tcp_async| Waiting for client connection at: " << acceptor.local_endpoint().address() << ":" << acceptor.local_endpoint().port() <<  std::endl;
+    acceptor.listen();
+
     Accept();
   }
 
   void OpcTcpServer::Shutdown()
   {
     std::clog << "opc_tcp_async| Shutting down server." << std::endl;
-    Clients.clear();
     acceptor.close();
+
+    // Actively shutdown OpcTcpConnections to clear open async requests from worker
+    // thread.
+    // Warning: the Clients container may be modified by OpcTcpConnections::GoodBye
+    // when calling Stop() which makes our iterator invalid. So have a copy of thiss
+    // container to iterate have a stable iterator.
+    Mutex.lock();
+    std::set<std::shared_ptr<OpcTcpConnection>> tmp(Clients);
+    Mutex.unlock();
+    auto i = tmp.begin();
+    while (i != tmp.end()) {
+      (*i)->Stop();
+      ++i;
+    }
+
+    // clear possibly remaining Client's
+
+    { std::unique_lock<std::mutex> lock(Mutex);
+      Clients.clear();
+    }
   }
 
   void OpcTcpServer::Accept()
   {
     try
     {
-      std::cout << "opc_tcp_async| Waiting for client connection at: " << acceptor.local_endpoint().address() << ":" << acceptor.local_endpoint().port() <<  std::endl;
-      acceptor.listen();
-      acceptor.async_accept(socket, [this](boost::system::error_code errorCode){
+      acceptor.async_accept(socket, [this](boost::system::error_code errorCode) {
+        if (!acceptor.is_open()) {
+            return;
```

Comment:
In OpcTcpServer::Shutdown() we do acceptor.close(). If shortly before we call close() the acceptor got a new connection request, this request may already be pushed to the async_io event loop but not yet be processed. So we can have a pending Accept() operation on an already closed acceptor. Thinking about it, it would be smart to wait for completion of pushed operations of the acceptor too, before Shutdown() is finished, because we may still run into a segfault when we get our very first connection request immediately before we close the acceptor. I will add that code in a later commit.